### PR TITLE
Runtime Implementation

### PR DESCRIPTION
**Pull request by ec2**
**Summary of changes**
Changes introduced in this pull request:
- implemented runtime
- Change interfaces for actors to take mutable references
- naive interpreter

**Reference issue to close (if applicable)**

Closes 
#67 

**Other information and links**

There are still a few things that need to be done. Namely, we still need to implement gas calculations. Until we have that, there we obviously cannot calculate the gas cost of calls. Also, we won't be able to verify some of the things that need to be verified. I have marked these as TODO. 

Also, I have left the Runtime package in because that is where the Runtime trait lives. It is debatable as to whether we need a Runtime trait and I am heavily leaning on removing it altogether. The initial reason we had the Runtime trait is to create different Runtime for testing, for example. However, we can mock those, which might actually make things a lot more easier. I figured that I'll leave the trait in here meanwhile, and possibly remove it when we start doing the Actors testing... But as an FYI, the VM is currently using the DefaultRuntime struct instead of referencing the trait. It is a LOT more easier to do it like this... 

The logic for the Lotus' send functions are in runtime.go and vm.go here https://github.com/filecoin-project/lotus/tree/testnet/3/chain/vm
@austinabell @dutterbutter

### CODE REVIEW

File: vm/default_runtime/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,442 @@
...
+                x if x == *MULTISIG_ACTOR_CODE_ID => {
+                    actor::cron::Actor.invoke_method(&mut *runtime, *method_num, msg.params())
+                }
+                x if x == *REWARD_ACTOR_CODE_ID => {
+                    actor::cron::Actor.invoke_method(&mut *runtime, *method_num, msg.params())
+                }
+                _ => todo!("Handle unknown code cids"),
+            }
+        };
+        return ret;
```
Comment (austinabell) [inline vm/default_runtime/src/lib.rs:401]: This does not revert state tree if failed

File: vm/default_runtime/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,442 @@
...
+        .ok_or_else(|| "transfer failed could not retrieve from actor".to_owned())?;
+    let mut to_actor = state
+        .get_actor(&to)
+        .map_err(|e| format!("transfer failed: {}", e))?
+        .ok_or_else(|| "transfer failed could not retrieve from actor".to_owned())?;
+
+    deduct_funds(&mut from_actor, &value)?;
+    deposit_funds(&mut to_actor, &value);
+    Ok(())
+}
```
Comment (austinabell) [inline vm/default_runtime/src/lib.rs:429]: Where are these changes persisted? You're changing the local copy, but this does nothing

File: vm/default_runtime/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,442 @@
...
+
+/// Safely deducts funds from an Actor
+fn deduct_funds(from: &mut ActorState, amt: &TokenAmount) -> Result<(), String> {
```
Comment (austinabell) [inline vm/default_runtime/src/lib.rs:432]:
```suggestion
fn deduct_funds(from: &mut ActorState, amt: &TokenAmount) -> Result<(), &'static str> {
```

File: vm/default_runtime/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,442 @@
...
+    if &from.balance < amt {
+        return Err("not enough funds".to_owned());
```
Comment (austinabell) [inline vm/default_runtime/src/lib.rs:434]:
```suggestion
        return Err("not enough funds");
```

File: vm/runtime/Cargo.toml [hunk 5]
```diff
@@ -9,6 +9,7 @@ vm = { path = "../../vm" }
 crypto = { path = "../../crypto" }
 address = { package = "forest_address", path = "../address" }
 message = { package = "forest_message", path = "../message" }
+chain = { path = "../../blockchain/chain" }
```
Comment (austinabell) [inline vm/runtime/Cargo.toml:12]:
```suggestion
```

File: vm/actor/src/builtin/multisig/mod.rs [hunk 6]
```diff
@@ -42,12 +42,13 @@ impl Method {
...
     /// Constructor for Multisig actor
-    pub fn constructor<BS, RT>(rt: &RT, params: ConstructorParams) -> Result<(), ActorError>
+    pub fn constructor<BS, RT>(rt: &mut RT, params: ConstructorParams) -> Result<(), ActorError>
     where
         BS: BlockStore,
         RT: Runtime<BS>,
     {
-        rt.validate_immediate_caller_is(std::iter::once(&*INIT_ACTOR_ADDR));
+        let sys_ref: &Address = &INIT_ACTOR_ADDR;
+        rt.validate_immediate_caller_is(std::iter::once(sys_ref))?;
```
[the remark this answers is not in the document]
Comment (ec2) [reply vm/actor/src/builtin/multisig/mod.rs]: fixed

File: vm/actor/src/builtin/init/mod.rs [hunk 7]
```diff
@@ -85,25 +86,25 @@ impl Actor {
...
-        rt.create_actor(&params.code_cid, &id_address);
+        rt.create_actor(&params.code_cid, &id_address)?;
 
         // Invoke constructor
         rt.send(
             &id_address,
             METHOD_CONSTRUCTOR,
             &params.constructor_params,
-            rt.message().value(),
+            rt.message().clone().value(),
```
[the remark this answers is not in the document]
Comment (ec2) [reply vm/actor/src/builtin/init/mod.rs]: fixed

File: vm/actor/src/builtin/paych/mod.rs [hunk 8]
```diff
@@ -40,63 +41,61 @@ impl Method {
...
-        if code_cid != *ACCOUNT_ACTOR_CODE_ID {
-            Err(format!(
-                "actor {} must be an account ({}), was {}",
-                raw, *ACCOUNT_ACTOR_CODE_ID, code_cid
+        let resolved = rt.resolve_address(raw)?;
+
+        let code_cid = rt.get_actor_code_cid(&resolved)?;
+
+        let account_code_ref: &Cid = &ACCOUNT_ACTOR_CODE_ID;
+        if &code_cid != account_code_ref {
```
[the remark this answers is not in the document]
Comment (ec2) [reply vm/actor/src/builtin/paych/mod.rs]: fixed

File: vm/actor/src/builtin/paych/mod.rs [hunk 9]
```diff
@@ -246,37 +249,42 @@ impl Actor {
...
-    pub fn collect<BS, RT>(rt: &RT) -> Result<(), ActorError>
+    pub fn collect<BS, RT>(rt: &mut RT) -> Result<(), ActorError>
     where
         BS: BlockStore,
         RT: Runtime<BS>,
     {
-        let st: State = rt.state();
-        rt.validate_immediate_caller_is([st.from.clone(), st.to.clone()].iter());
+        let st: State = rt.state()?;
+        rt.validate_immediate_caller_is([st.from.clone(), st.to.clone()].iter())?;
```
[the remark this answers is not in the document]
Comment (ec2) [reply vm/actor/src/builtin/paych/mod.rs]: fixed

File: vm/actor/src/builtin/power/mod.rs [hunk 10]
```diff
@@ -170,22 +160,19 @@ impl Actor {
...
-                &addresses.id_address,
-                rt.message().value().clone(),
-            )
-            .map_err(|e| {
-                rt.abort(
-                    ExitCode::ErrIllegalState,
-                    format!("failed to set pledge balance {}", e),
-                )
-            })?;
+        let msg = rt.message().value().clone();
```
[the remark this answers is not in the document]
Comment (ec2) [reply vm/actor/src/builtin/power/mod.rs]: fixed

File: vm/actor/src/builtin/system/mod.rs [hunk 11]
```diff
@@ -32,7 +33,8 @@ impl Actor {
         BS: BlockStore,
         RT: Runtime<BS>,
     {
-        rt.validate_immediate_caller_is(std::iter::once(&*SYSTEM_ACTOR_ADDR));
+        let sys_ref: &Address = &SYSTEM_ACTOR_ADDR;
+        rt.validate_immediate_caller_is(std::iter::once(sys_ref))?;
```
[the remark this answers is not in the document]
Comment (ec2) [reply vm/actor/src/builtin/system/mod.rs]: fixed

File: vm/default_runtime/Cargo.toml [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+authors = ["ChainSafe Systems <info@chainsafe.io>"]
+edition = "2018"
+
+[dependencies]
+address = { package = "forest_address", path = "../address" }
+actor = { path = "../actor" }
+message = { package = "forest_message", path = "../message" }
+runtime = { path = "../runtime" }
+blocks = { package = "forest_blocks", path = "../../blockchain/blocks" }
+chain = { path = "../../blockchain/chain" }
```
[the remark this answers is not in the document]
Comment (ec2) [reply vm/default_runtime/Cargo.toml]: fixed

File: vm/default_runtime/src/lib.rs [hunk 13]
```diff
@@ -0,0 +1,442 @@
...
+
+impl<'a, 'b, 'c, ST: StateTree, BS: BlockStore> DefaultRuntime<'a, 'b, 'c, ST, BS> {
+    /// Constructs a new Runtime
+    pub fn new(
+        state: &'c mut ST,
+        store: &'a BS,
+        gas_used: u64,
+        message: &'b UnsignedMessage,
+        epoch: ChainEpoch,
+        origin: &Address,
```
[the remark this answers is not in the document]
Comment (ec2) [reply vm/default_runtime/src/lib.rs]: fixed

File: vm/default_runtime/src/lib.rs [hunk 14]
```diff
@@ -0,0 +1,442 @@
...
+        origin_nonce: u64,
+    ) -> Self {
+        DefaultRuntime {
+            state,
+            store,
+            gas_used,
+            gas_available: message.gas_limit(),
+            message,
+            epoch,
+            origin: origin.clone(),
```
[the remark this answers is not in the document]
Comment (ec2) [reply vm/default_runtime/src/lib.rs]: fixed

File: vm/default_runtime/src/lib.rs [hunk 15]
```diff
@@ -0,0 +1,442 @@
...
+    fn get_randomness(
+        _personalization: DomainSeparationTag,
+        _rand_epoch: ChainEpoch,
+        _entropy: &[u8],
+    ) -> Randomness {
+        todo!()
+    }
+
+    fn create<C: Cbor>(&mut self, obj: &C) -> Result<(), ActorError> {
+        // TODO: Verify if right hash to use
```
[the remark this answers is not in the document]
Comment (ec2) [reply vm/default_runtime/src/lib.rs]: fixed

File: vm/default_runtime/src/lib.rs [hunk 16]
```diff
@@ -0,0 +1,442 @@
...
+                self.abort(
+                    ExitCode::ErrPlaceholder,
+                    format!("storage get error in read only state: {}", e.to_string()),
+                )
+            })
+            .and_then(|c| {
+                c.ok_or_else(|| {
+                    self.abort(
+                        ExitCode::ErrPlaceholder,
+                        "storage get error in  read only state".to_owned(),
```
[the remark this answers is not in the document]
Comment (ec2) [reply vm/default_runtime/src/lib.rs]: fixed

File: vm/default_runtime/src/lib.rs [hunk 17]
```diff
@@ -0,0 +1,442 @@
...
+                    )
+                })
+            })
```
[the remark this answers is not in the document]
Comment (ec2) [reply vm/default_runtime/src/lib.rs]: fixed. Hmm, i think my brain kinda went off somewhere while writing that

File: vm/default_runtime/src/lib.rs [hunk 18]
```diff
@@ -0,0 +1,442 @@
...
+                )
+            })
+            .and_then(|c| {
+                c.ok_or_else(|| {
+                    self.abort(
+                        ExitCode::ErrPlaceholder,
+                        "storage get error in transaction".to_owned(),
+                    )
+                })
+            })?;
```
[the remark this answers is not in the document]
Comment (ec2) [reply vm/default_runtime/src/lib.rs]: fixed

File: vm/default_runtime/src/lib.rs [hunk 19]
```diff
@@ -0,0 +1,442 @@
...
+        if send_res.is_err() {
+            self.state
+                .revert_to_snapshot(&snapshot)
+                .map_err(|_e| self.abort(ExitCode::ErrPlaceholder, "failed to revert snapshot"))?;
+        }
+        send_res
+    }
+
+    fn abort<S: AsRef<str>>(&self, _exit_code: ExitCode, _msg: S) -> ActorError {
+        todo!()
```
[the remark this answers is not in the document]
Comment (ec2) [reply vm/default_runtime/src/lib.rs]: fixed

File: vm/default_runtime/src/lib.rs (hunk 2 above)
Comment (ec2) [reply vm/default_runtime/src/lib.rs]: fixed

File: vm/runtime/Cargo.toml (hunk 5 above)
Comment (ec2) [reply vm/runtime/Cargo.toml:12]: fixed

File: vm/default_runtime/src/lib.rs [hunk 20]
```diff
@@ -0,0 +1,442 @@
...
+            )
+        })
+        .and_then(|act| {
+            act.ok_or_else(|| {
+                ActorError::new(
+                    ExitCode::SysErrInternal,
+                    "actor not found in internal_send".to_owned(),
+                )
+            })
+        })?;
```
[the remark this answers is not in the document]
Comment (ec2) [reply vm/default_runtime/src/lib.rs:344]: fixed

**Pull request merged**